### PR TITLE
build: set -DNDEBUG in Release

### LINKED ISSUE #22745: NDEBUG is not set in Release mode
## 🐛 Bug

The generated VariableType has debug code that looks like:

```
  #ifndef NDEBUG
  c10::optional<Storage> self_storage_saved =
    self.has_storage() ? c10::optional<Storage>(self.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self_impl_saved;
  if (self.defined()) self_impl_saved = self.getIntrusivePtr();
  c10::optional<Storage> other_storage_saved =
    other.has_storage() ? c10::optional<Storage>(other.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> other_impl_saved;
  if (other.defined()) other_impl_saved = other.getIntrusivePtr();
  #endif
```

We don't define `NDEBUG` in Release mode so this code always runs. It adds ~100 ns to the overhead of every operation. 

It's generated by:
 https://github.com/pytorch/pytorch/blob/89d6e8804267d467e39529d710e2da39982905fa/tools/autograd/gen_variable_type.py#L263-L267

cc @ezyang @gchanan @zou3519 @yf225

### PR DESCRIPTION
**Pull request by Baranowski**
This might lead to silent undefined behaviour (e.g. with out-of-bound indices). This affects `test_multinomial_invalid_probs_cuda` which is now removed.

Test plan:
* Build with VERBOSE=1 and manually inspect `less ndebug.build.log | grep 'c++' | grep -v -- -DNDEBUG` (only with nina on Linux)
* CI

Fixes #22745

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit c04fc3fb:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c04fc3fbb28b28fa7bdf852d1b471980f6d65516).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4364686) pytorch_linux_xenial_cuda10_1_cudnn7_py3_slow_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92280049) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92280049))
<details>
<summary>
<code>Jan 29 00:29:25 RuntimeError: test_cuda failed!</code>
</summary>

```
Jan 29 00:29:24 Ran 98 tests in 5.167s 
Jan 29 00:29:24  
Jan 29 00:29:24 FAILED (failures=1, skipped=97) 
Jan 29 00:29:24  
Jan 29 00:29:24 Generating XML reports... 
Jan 29 00:29:25 Traceback (most recent call last): 
Jan 29 00:29:25   File "test/run_test.py", line 456, in <module> 
Jan 29 00:29:25     main() 
Jan 29 00:29:25   File "test/run_test.py", line 449, in main 
Jan 29 00:29:25     raise RuntimeError(message) 
Jan 29 00:29:25 RuntimeError: test_cuda failed! 
Jan 29 00:29:25 + cleanup 
Jan 29 00:29:25 + retcode=1 
Jan 29 00:29:25 + set +x 
Jan 29 00:29:25 =================== sccache compilation log =================== 
Jan 29 00:29:25 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Jan 29 00:29:25  
Jan 29 00:29:25 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 29 00:29:25 Compile requests               145 
Jan 29 00:29:25 Compile requests executed       52 
Jan 29 00:29:25 Cache hits                      40 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by Baranowski** [comment]
(note to self) Currently there is only one test failing: `test_multinomial_invalid_probs_cuda` in `ci/circleci: pytorch_linux_xenial_cuda10_1_cudnn7_py3_slow_test`. I will try to trigger another run of CI to see if this persists.

**Comment by Baranowski** [comment]
I feel that removing `test_multinomial_invalid_probs_cuda` is the wrong way to go about it. It would make sense to define a new constant or function like `torch.compiled_with_safety_assertions()` and disable the test only if the function returns false. But I didn't want to do that until someone confirms that it makes sense.

**Comment by yf225** [comment]
@Baranowski Would you like to try whether the changes in `aten/src/THC/THCTensorRandom.cuh` in https://github.com/pytorch/pytorch/pull/16827/files would fix the `test_multinomial_invalid_probs_cuda` issue? Also I believe we should address https://github.com/pytorch/pytorch/pull/16827#issuecomment-461238475.

File: cmake/Dependencies.cmake
Comment (peterjc123) [inline cmake/Dependencies.cmake]: This won't work for multiple-build-type generators. For better solutions, please refer to https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#make-override-files and https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#dynamic-replace.

**Comment by Baranowski** [comment]
@yf225 You were right, that patch fixed `test_multinomial_invalid_probs_cuda`

File: cmake/Dependencies.cmake
Comment (Baranowski) [inline cmake/Dependencies.cmake]: @peterjc123 Based on my reading of the docs, I think that this should take care of the multi-build-type generators since they should use `CMAKE_CXX_FLAGS_RELEASE`. Unfortunately, I work only on Linux and have no way to test this for MSVC or other generators. Perhaps I should stop working on this and leave it up to someone with more diverse environments?

Comment (peterjc123) [reply cmake/Dependencies.cmake]: Actually, it's acceptable because we don't  use the MSBuild generator in our MSVC build. Alternatively, we use Ninja which is a single-build-type generator. Maybe we can open a new issue for this and skip that for now.

Comment (Baranowski) [reply cmake/Dependencies.cmake]: @peterjc123 Just to make sure I understand you correctly, you are suggesting that I:
* revert the change of the `*_RELEASE` flags
* create a new issue to support MSBuild and possibly other multi-build-type generators. (If they are not being supported now, I suspect there will be more discrepancies than just NDEBUG).

Comment (peterjc123) [reply cmake/Dependencies.cmake]: May I know what you are doing first? I'm a bit confused about the changes here. The code here is to remove `-DNDEBUG` from the compiler flags from both debug and release builds if `CMAKE_BUILD_TYPE` is `Release`.

Comment (Baranowski) [reply cmake/Dependencies.cmake]: Sorry, changing `*_DEBUG` flags if build type is Release is indeed confusing. I will clean up that part and submit a clearer patch, even if only for the sake of discussion.

File: cmake/Dependencies.cmake
Comment (Baranowski) [inline cmake/Dependencies.cmake]: @peterjc123 Here is the new proposed version. I removed the `OPTION(NDEBUG ...)` since I don't see much use for it after `NDEBUG` is automatically set/unset depending on the build type

**Review by peterjc123 — APPROVED** [review]
Sorry for late reply. This one looks much better. lgtm

**Comment by Baranowski** [comment]
No worries, @peterjc123. I have updated the PR description to reflect that I only tested on ninja+Linux

**Comment by yf225** [comment]
It seems that the changes in `aten/src/ATen/native/cuda/MultinomialKernel.cu` is causing the build to fail. @smessmer Curious do you know whether we can use `AT_ASSERT` or similar macros in device code?

**Comment by Baranowski** [comment]
Sorry, yf255 is correct. I had only checked the kostmo report. The failing builds are:

- pytorch_libtorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build\
- pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build
- pytorch_windows_build

And the error msgs are all along the lines:
```
Feb 01 10:26:57 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/MultinomialKernel.cu(243): error: device code does not support exception handling
Feb 01 10:26:57 
Feb 01 10:26:57 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/MultinomialKernel.cu(244): error: device code does not support exception handling
Feb 01 10:26:57 
Feb 01 10:26:57 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/MultinomialKernel.cu(245): error: device code does not support exception handling
```

To answer @yf225 's question, I'm sure there is a way to assert in the device code, since `assert(...)` with **no** `-DNDEBUG` flag was compiling and reporting the error as expected. It's still not clear to me whether that's the standard `<cassert>` macro or a custom definition. Will dig in and fix the builds.

**Comment by Baranowski** [comment]
This `assert()` seems to be defined in the CUDA header `crt/common_functions.h` (`#if defined(__CUDA_RTC__)`). I'm now trying to add a macro in `c10/util/Exception.h` with similar functionality but ignoring `NDEBUG`.

File: c10/macros/Macros.h
Comment (peterjc123) [inline c10/macros/Macros.h]: In MSVC, the function is named as `_wassert` and you should declare it as the following.
```c++
_declspec(dllimport) void _wassert(
	wchar_t const* _Message,
	wchar_t const* _File,
	unsigned       _Line
);

#define __assert_fail(expression) (void)(                                                 \
        (!!(expression)) ||                                                               \
        (_wassert(_CRT_WIDE(#expression), _CRT_WIDE(__FILE__), (unsigned)(__LINE__)), 0)  \
    )
```
Actually, this is copied from `assert.h` in MSVC.

Comment (Baranowski) [reply c10/macros/Macros.h]: Sorry for putting it on hold for the past week.

Thanks for the hint. Unfortunately, I have no way to test this, and so I would rather not commit it. I propose that I will simply disable CUDA_ALWAYS_ASSERT for MSVC builds, comment with a TODO and create a new issue to implement it later.

The tests should pass nevertheless since the one test affected is already disabled on Windows (`skipIf(IS_WINDOWS, 'FIXME: CUDA OOM error on Windows')`).

Comment (peterjc123) [reply c10/macros/Macros.h]: Ok, no problem for me. I'll fix that for you after this PR goes in.

**Comment by Baranowski** [comment]
I think this is ready. To my untrained eyes, none of the CI failures look related.

**Comment by Baranowski** [comment]
@peterjc123, is there anything else left for me to do here?

**Comment by peterjc123** [comment]
Nope. @yf225 Would you please take a look?

**Comment by peterjc123** [comment]
BTW, you may need to rebase this PR on master.

**Review by yf225 — APPROVED** [review]
@Baranowski Thanks a lot for the awesome work!

**Pull request closed without merging**
